### ISSUE TITLE
remote pac scrips and proxy

### ISSUE THREAD

**Issue by jeka** [FelisCatus/SwitchyOmega#910]
Hi,

Several times I have observed that my pac script doesnt look as it should be. My guess is that SwitchyOmega downloads pac scripts set in PAC profiles using current proxy settings. And because my company is using dynamic pac file which is generated on request and depends on IP from which requests is coming things get messed up.
Now I am not sure is there a valid use case of using proxy for downloading pac scripts but I would suggest to ether force direct connection for this task or at least some way of configuring the behavior.

And of course there is a possibility that I am completely wrong and its not because of SwitchyOmega at all.

**Comment by FelisCatus** [comment]
SwitchyOmega is using AJAX like the rest of the web and must follow the proxy rules when downloading stuff. Absolutely no control over this.

On way to work around this would be passing the URL directly to Chrome settings. When Chrome does the proxy resolving routine, it will connect directly on its own. However, that would make the PAC Profile unusable in Auto Switch and other profiles because SwitchyOmega never knows the exact content of the script and cannot craft it into the master PAC file. Currently, this method is only used for PAC files on the local file system because SO cannot read files on the disk. (You can try to change the URL to `file:///home/user/a.pac` to see the message shown for this.)

Another way to do this requires more user configuration: Use a Switch Profile to assign DIRECT to the particular URL of the PAC. The configuration would essentially look like this:

* Switch Profile
  * `http://wpad/wpad.dat` (or whatever rule) => `[Direct Connection]`
  * `(default)` => `PAC Profile`

.. and keep that Switch Profile activated all the time in the dropdown menu.

Besides, I'm totally fine with adding an option and documenting the limitations on the settings page. Something like: [ ] Always use DIRECT connection when downloading the PAC script. And a limitation message will be shown when checked.

**Comment by jeka** [comment]
I see ... I already have such rule (DIRECT for pac script) in auto-switch mode. The problem is that I regularly have to switch to "hard" proxy profile and I forget to switch back sometimes.
I know that I can force specific profile on startup but that would be exactly as much annoying as messed up proxy script. Another option is to play with pac script download frequency but this is global option and I would prefer it to be per profile (I have 3 different pac profiles in auto-switch mode). 
Making PAC Profiles unusable in auto-switch mode is a HUGE drawback and its not what I am asking at all. I am using them there extensively.
And about pac scripts on local filesystem. I guess I have seen the message in the past because I am using thttpd locally just for serving pac scripts.

**Comment by FelisCatus** [comment]
If any options are provided, I will certainly make it local to each profile. I totally understand.

**Comment by FelisCatus** [comment]
Or we can add a implied rule to all profiles when applied to use direct connection for the PAC URL.

In this way, no matter what profile you are on (except System Profile and PAC Profile with local files), the PAC URL always use direct connection. What do you think?

**Comment by jeka** [comment]
I don't understand what do you propose actually.

I realized that I can use 'Bypass' list on proxy profiles to workaround this issue.
My problem is that I have tons of such profiles and I set temporal ones up on daily basis. Too much effort to keep them updated :/

So I was thinking about something like global  'Bypass' list which is applied to all proxy profiles.
Or even better let profile chose to use global list or not (in additional to local).

**Comment by FelisCatus** [comment]
Yeah, that basically what I meant technically. But I didn't think of making it so flexible though -- I proposed that checking that option for a PAC Profile should place its URL into the hidden global list. In this way we can achieve the same thing with a simpler user interface.

For example: suppose we have Proxy 1, Proxy 2 and PAC Profile 3.

In PAC Profile 3, there is an option that says "always use direct connection when downloading the script." When it is checked, the following logic applies:

* When you activate Proxy 1, the PAC URL is effectively inserted to the bypass list of Proxy 1.
  - No actual changes are made, the rule is internally enforced but hidden to the user.
* When you activate Proxy 2, the PAC URL is effectively inserted to the bypass list of Proxy 2.
* When you activate a Switch Profile, the PAC URL is effectively inserted as the first rule in the Switch Profile.
* and so on...
* But when you use System Proxy, no global rule applies because SwitchyOmega has no control over that.
  - This is also the case with other PAC Profiles using a local file.

Am I making sense?

**Comment by jeka** [comment]
Ah yes, I understand now, that is better idea. Simplicity is a GoodThing<sup>:tm:</sup>. 
Just one note: I have multiple PAC profiles so this hidden global bypass list should have more than 1 entry.

**Comment by FelisCatus** [comment]
Great, I will try to do this later. I think we need pretty good documentation/help text about the limitations here though.
